# Data science in Python

- Course GitHub repo: https://github.com/pycam/python-data-science
- Python website: https://www.python.org/ 

## Session 2.1: Working with Pandas

- [Reading CSV Data Using Pandas](#Reading-CSV-Data-Using-Pandas)
- [Exploring our data](#Exploring-our-data)
- [DataFrame manipulation](#DataFrame-manipulation)
- [Exercise 2.1.1](#Exercise-2.1.1)

## Mind map

<img src="img/mind_maps/mind_maps.004.jpeg">

## Reading CSV Data Using Pandas

### Import the Pandas library

Pandas is a widely-used external Python library for statistics, particularly on tabular data.
It borrows many features from R’s dataframes.
A dataframe is a 2-dimentional table whose columns have names and potentially have different data types.

Pandas website http://pandas.pydata.org/ and documentation http://pandas.pydata.org/pandas-docs/stable/.

To load `pandas` into your environment, you first need to install it using `pip install pandas` as it is an external third-party library, it is not included by default when you install Python.

When installed, to load it, use `import pandas`:

In [2]:
import pandas

### Read CSV data

For reading a Comma Separate Values (CSV) data file with pandas, we use `pandas.read_csv()`:

- Argument is the name of the file to be read.
- Assign result to a variable to store the data that was read.


The columns in a dataframe are the observed variables, and the rows are the observations. We are going to load a slightly different Gapminder dataset for Oceania, where each columns represent the GDP per capita on different years and each rows a country in Oceania. 

Pandas uses backslash `\` to show wrapped lines when output is too wide to fit the screen.

In [6]:
data = pandas.read_csv('data/gapminder_gdp_oceania.csv')
print(data)

       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Australia     10039.59564     10949.64959     12217.22686   
1  New Zealand     10556.57566     12247.39532     13175.67800   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     14526.12465     16788.62948     18334.19751     19477.00928   
1     14463.91893     16046.03728     16233.71770     17632.41040   

   gdpPercap_1987  gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  \
0     21888.88903     23424.76683     26997.93657     30687.75473   
1     19007.19129     18363.32494     21050.41377     23189.80135   

   gdpPercap_2007  
0     34435.36744  
1     25185.00911  


In [5]:
data = pandas.read_csv('data/gapminder_gdp_oceania.csv', index_col='country')
print(data)

             gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
country                                                                       
Australia       10039.59564     10949.64959     12217.22686     14526.12465   
New Zealand     10556.57566     12247.39532     13175.67800     14463.91893   

             gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                       
Australia       16788.62948     18334.19751     19477.00928     21888.88903   
New Zealand     16046.03728     16233.71770     17632.41040     19007.19129   

             gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  
country                                                                      
Australia       23424.76683     26997.93657     30687.75473     34435.36744  
New Zealand     18363.32494     21050.41377     23189.80135     25185.00911  


Our course stores its data files in a `data/` sub-directory, which is why the path to the file is `data/gapminder_gdp_oceania.csv`. If you forget to include `data/`, or if you include it but your copy of the file is somewhere else, you will get a runtime error that ends with a line like this:
```
FileNotFoundError: File b'gapminder_gdp_oceania.csv' does not exist
```

## Exploring our data

A DataFrame is a 2-dimensional data structure that can store data of different types (including characters, integers, floating point values, factors and more) in columns. It is similar to a spreadsheet or an SQL table or the `data.frame` in R. A DataFrame always has an index (0-based). An index refers to the position of an element in the data structure.

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 13 columns):
country           2 non-null object
gdpPercap_1952    2 non-null float64
gdpPercap_1957    2 non-null float64
gdpPercap_1962    2 non-null float64
gdpPercap_1967    2 non-null float64
gdpPercap_1972    2 non-null float64
gdpPercap_1977    2 non-null float64
gdpPercap_1982    2 non-null float64
gdpPercap_1987    2 non-null float64
gdpPercap_1992    2 non-null float64
gdpPercap_1997    2 non-null float64
gdpPercap_2002    2 non-null float64
gdpPercap_2007    2 non-null float64
dtypes: float64(12), object(1)
memory usage: 288.0+ bytes


As expected, it’s a `DataFrame` (or, to use the full name that Python uses to refer to it internally, a `pandas.core.frame.DataFrame`).

It has 2 rows and 13 columns. It uses 288 bytes of memory.

The row headings are numbers (0 and 1 in this case) but we really want to index this DataFrame by country. To do so, we pass the name of the column to `read_csv()` as its `index_col` parameter to do this:

In [8]:
data = pandas.read_csv('data/gapminder_gdp_oceania.csv', index_col='country')
print(data)

             gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
country                                                                       
Australia       10039.59564     10949.64959     12217.22686     14526.12465   
New Zealand     10556.57566     12247.39532     13175.67800     14463.91893   

             gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                       
Australia       16788.62948     18334.19751     19477.00928     21888.88903   
New Zealand     16046.03728     16233.71770     17632.41040     19007.19129   

             gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  
country                                                                      
Australia       23424.76683     26997.93657     30687.75473     34435.36744  
New Zealand     18363.32494     21050.41377     23189.80135     25185.00911  


In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2 entries, Australia to New Zealand
Data columns (total 12 columns):
gdpPercap_1952    2 non-null float64
gdpPercap_1957    2 non-null float64
gdpPercap_1962    2 non-null float64
gdpPercap_1967    2 non-null float64
gdpPercap_1972    2 non-null float64
gdpPercap_1977    2 non-null float64
gdpPercap_1982    2 non-null float64
gdpPercap_1987    2 non-null float64
gdpPercap_1992    2 non-null float64
gdpPercap_1997    2 non-null float64
gdpPercap_2002    2 non-null float64
gdpPercap_2007    2 non-null float64
dtypes: float64(12)
memory usage: 208.0+ bytes


We can also use the `type()` function to see what kind of thing data is:

In [12]:
type(data)

pandas.core.frame.DataFrame

To see what kind of things does data contain, DataFrames have an attribute called dtypes which returns the data type of each columns. Note that this is an attribute associated to the DataFrame data, and not a method. So do not use `()` to call it.

In [11]:
data.dtypes

gdpPercap_1952    float64
gdpPercap_1957    float64
gdpPercap_1962    float64
gdpPercap_1967    float64
gdpPercap_1972    float64
gdpPercap_1977    float64
gdpPercap_1982    float64
gdpPercap_1987    float64
gdpPercap_1992    float64
gdpPercap_1997    float64
gdpPercap_2002    float64
gdpPercap_2007    float64
dtype: object

It has now 2 rows named 'Australia' and 'New Zealand' and 12 columns, each of which has two actual 64-bit floating point values. It uses 208 bytes of memory.

There are many ways to summarize and access the data stored in DataFrames, using attributes and methods provided by the [DataFrame object](https://pandas.pydata.org/pandas-docs/stable/api.html#dataframe).

To access an attribute, use the DataFrame object name followed by the attribute name `df_object.attribute`. Using the DataFrame `data` and attribute `columns`, an index of all the column names in the DataFrame can be accessed with `data.columns`.

Methods are called in a similar fashion using the syntax `df_object.method()`. As an example, `data.head()` gets the first few rows in the DataFrame `data` using the `head()` method. With a method, we can supply extra information in the parenthesis as arguments to control behaviour.

Let’s look at the data using these.

In [13]:
data.columns

Index(['gdpPercap_1952', 'gdpPercap_1957', 'gdpPercap_1962', 'gdpPercap_1967',
       'gdpPercap_1972', 'gdpPercap_1977', 'gdpPercap_1982', 'gdpPercap_1987',
       'gdpPercap_1992', 'gdpPercap_1997', 'gdpPercap_2002', 'gdpPercap_2007'],
      dtype='object')

Let's load European's data to have more rows:

In [14]:
eu_data = pandas.read_csv('data/gapminder_gdp_europe.csv', index_col='country')
print(eu_data)

                        gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                                  
Albania                    1601.056136     1942.284244     2312.888958   
Austria                    6137.076492     8842.598030    10750.721110   
Belgium                    8343.105127     9714.960623    10991.206760   
Bosnia and Herzegovina      973.533195     1353.989176     1709.683679   
Bulgaria                   2444.286648     3008.670727     4254.337839   
Croatia                    3119.236520     4338.231617     5477.890018   
Czech Republic             6876.140250     8256.343918    10136.867130   
Denmark                    9692.385245    11099.659350    13583.313510   
Finland                    6424.519071     7545.415386     9371.842561   
France                     7029.809327     8662.834898    10560.485530   
Germany                    7144.114393    10187.826650    12902.462910   
Greece                     3530.690067

In [15]:
eu_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30 entries, Albania to United Kingdom
Data columns (total 12 columns):
gdpPercap_1952    30 non-null float64
gdpPercap_1957    30 non-null float64
gdpPercap_1962    30 non-null float64
gdpPercap_1967    30 non-null float64
gdpPercap_1972    30 non-null float64
gdpPercap_1977    30 non-null float64
gdpPercap_1982    30 non-null float64
gdpPercap_1987    30 non-null float64
gdpPercap_1992    30 non-null float64
gdpPercap_1997    30 non-null float64
gdpPercap_2002    30 non-null float64
gdpPercap_2007    30 non-null float64
dtypes: float64(12)
memory usage: 3.0+ KB


Let's find out what the head() method does:

In [16]:
help(eu_data.head)

Help on method head in module pandas.core.generic:

head(n=5) method of pandas.core.frame.DataFrame instance
    Return the first n rows.
    
    Parameters
    ----------
    n : int, default 5
        Number of rows to select.
    
    Returns
    -------
    obj_head : type of caller
        The first n rows of the caller object.



In [19]:
eu_data.head()

gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                                  
Albania                    1601.056136     1942.284244     2312.888958   
Austria                    6137.076492     8842.598030    10750.721110   
Belgium                    8343.105127     9714.960623    10991.206760   
Bosnia and Herzegovina      973.533195     1353.989176     1709.683679   
Bulgaria                   2444.286648     3008.670727     4254.337839   

                        gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
country                                                                  
Albania                    2760.196931     3313.422188     3533.003910   
Austria                   12834.602400    16661.625600    19749.422300   
Belgium                   13149.041190    16672.143560    19117.974480   
Bosnia and Herzegovina     2172.352423     2860.169750     3528.481305   
Bulgaria                   5577.002800     6597.494398     7612.240438   

                        gdpPercap_1982  gdpPercap_1987  gdpPercap_1992  \
country                                                                  
Albania                    3630.880722     3738.932735     2497.437901   
Austria                   21597.083620    23687.826070    27042.018680   
Belgium                   20979.845890    22525.563080    25575.570690   
Bosnia and Herzegovina     4126.613157     4314.114757     2546.781445   
Bulgaria                   8224.191647     8239.854824     6302.623438   

                        gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  
country                                                                 
Albania                    3193.054604     4604.211737     5937.029526  
Austria                   29095.920660    32417.607690    36126.492700  
Belgium                   27561.196630    30485.883750    33692.605080  
Bosnia and Herzegovina     4766.355904     6018.975239     7446.298803  
Bulgaria                   5970.388760     7696.777725    10680.792820

In [ ]:
eu_data.head(3)

In [20]:
eu_data.describe()

gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
count       30.000000       30.000000       30.000000       30.000000   
mean      5661.057435     6963.012816     8365.486814    10143.823757   
std       3114.060493     3677.950146     4199.193906     4724.983889   
min        973.533195     1353.989176     1709.683679     2172.352423   
25%       3241.132406     4394.874315     5373.536612     6657.939047   
50%       5142.469716     6066.721495     7515.733738     9366.067033   
75%       7236.794919     9597.220820    10931.085347    13277.182057   
max      14734.232750    17909.489730    20431.092700    22966.144320   

       gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
count       30.000000       30.000000       30.000000       30.000000   
mean     12479.575246    14283.979110    15617.896551    17214.310727   
std       5509.691411     5874.464896     6453.234827     7482.957960   
min       2860.169750     3528.481305     3630.880722     3738.932735   
25%       9057.708095    10360.030300    11449.870115    12274.570680   
50%      12326.379990    14225.754515    15322.824720    16215.485895   
75%      16523.017127    19052.412163    20901.729730    23321.587723   
max      27195.113040    26982.290520    28397.715120    31540.974800   

       gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  
count       30.000000       30.000000       30.000000       30.000000  
mean     17061.568084    19076.781802    21711.732422    25054.481636  
std       9109.804361    10065.457716    11197.355517    11800.339811  
min       2497.437901     3193.054604     4604.211737     5937.029526  
25%       8667.113214     9946.599306    11721.851483    14811.898210  
50%      17550.155945    19596.498550    23674.863230    28054.065790  
75%      25034.243045    27189.530312    30373.363307    33817.962533  
max      33965.661150    41283.164330    44683.975250    49357.190170

We often want to calculate summary statistics grouped by subsets or attributes within fields of our data. For example, we might want to calculate the average GDP per capita for 1962.

We can calculate basic statistics for all records in a single column using the syntax below:

In [21]:
eu_data.gdpPercap_1962.describe()

count       30.000000
mean      8365.486814
std       4199.193906
min       1709.683679
25%       5373.536612
50%       7515.733738
75%      10931.085347
max      20431.092700
Name: gdpPercap_1962, dtype: float64

In [22]:
eu_data['gdpPercap_1962'].describe()

count       30.000000
mean      8365.486814
std       4199.193906
min       1709.683679
25%       5373.536612
50%       7515.733738
75%      10931.085347
max      20431.092700
Name: gdpPercap_1962, dtype: float64

We can also extract one specific metric if we wish:

In [ ]:
eu_data['gdpPercap_1962'].mean()

In [ ]:
eu_data['gdpPercap_1962'].std()

The pandas function `describe()` will return descriptive stats including: mean, median, max, min, std and count for a particular column in the data. Pandas’ describe function will only return summary values for columns containing numeric data.

It is not particularly useful with few records, but it could be very helpful when there are thousands.

## DataFrame manipulation

### Pandas Cheat Sheet
[Pandas Cheat Sheet](http://pandas.pydata.org/Pandas_Cheat_Sheet.pdf)

In [23]:
import pandas
df = pandas.DataFrame({'gene': ['BRCA2', 'TNFAIP3', 'TCF7'], 
                       'chrom': ['13', '6', '5'],
                       'len': [84195, 16099, 37155]}
)
print(df)

  chrom     gene    len
0    13    BRCA2  84195
1     6  TNFAIP3  16099
2     5     TCF7  37155


### Selecting columns and rows
Select just one column:

In [24]:
df['len']

0    84195
1    16099
2    37155
Name: len, dtype: int64

or the first two rows:

In [47]:
df[:2]

chrom     gene    len
0    13    BRCA2  84195
1     6  TNFAIP3  16099

We can combine these to get the first 2 rows of the column 'len':

In [48]:
df[:2]['len']

0    84195
1    16099
Name: len, dtype: int64

This can also be done using **positional indexing** and `.iloc[]`:

In [49]:
df.iloc[:2]

chrom     gene    len
0    13    BRCA2  84195
1     6  TNFAIP3  16099

In [50]:
df.iloc[:2,2]

0    84195
1    16099
Name: len, dtype: int64

or to get the first 2 rows of all columns from indexed 1 till the end:

In [55]:
df.iloc[:2,1:]

gene    len
0    BRCA2  84195
1  TNFAIP3  16099

We have been using positional indexing so far, but we can also use **label-based indexing** with `.loc[]`:

In [56]:
df.loc[:1,'chrom':]

chrom     gene    len
0    13    BRCA2  84195
1     6  TNFAIP3  16099

### Sorting columns

In [57]:
sorted_df = df.sort_values('len')
print(sorted_df)

  chrom     gene    len
1     6  TNFAIP3  16099
2     5     TCF7  37155
0    13    BRCA2  84195


In [58]:
sorted_df = df.sort_values('len', ascending=False)
print(sorted_df)

  chrom     gene    len
0    13    BRCA2  84195
2     5     TCF7  37155
1     6  TNFAIP3  16099


## Exercise 2.1.1

- Read the data in `gapminder_gdp_americas.csv` (which should be in the same directory as `gapminder_gdp_oceania.csv`) into a variable called `americas_data` and display its summary statistics.
- As well as the `read_csv()` function for reading data from a file, Pandas provides a `to_csv()` function to write dataframes to files. Applying what you’ve learned about reading from files, write one of your dataframes to a file called `processed.csv`. You can use help to get information on how to use `to_csv`.

In [ ]:
import pandas as pd

# read data in 
americas_data = pd.read_csv('data/gapminder_gdp_americas.csv')
# summary stats
americas_data.describe()
# output to csv
americas_data.to_csv('data/americas_test.csv', index = False)

In [109]:
import pandas

americas_data = pandas.read_csv('data/gapminder_gdp_americas.csv',index_col='country')
#americas_data.loc[:,'gdpPercap_1952':]
#americas_data['gdpPercap_1952'].describe()
#americas_data.to_csv('data/processed.csv')

#americas_data.head()
#americas_data['continent']
#americas_data[:2]['gdpPercap_1952']
#americas_data.loc[:'Brazil','gdpPercap_1957':]
#americas_data.iloc[:5,:5]

sorted_df = americas_data.iloc[:5,:5].sort_values('gdpPercap_1952', ascending = False)
print(sorted_df)

          continent  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                               
Canada     Americas    11367.161120    12489.950060    13462.485550   
Argentina  Americas     5911.315053     6856.856212     7133.166023   
Chile      Americas     3939.978789     4315.622723     4519.094331   
Bolivia    Americas     2677.326347     2127.686326     2180.972546   
Brazil     Americas     2108.944355     2487.365989     3336.585802   

           gdpPercap_1967  
country                    
Canada       16076.588030  
Argentina     8052.953021  
Chile         5106.654313  
Bolivia       2586.886053  
Brazil        3429.864357  


## Manipulating data with Pandas (live coding session)

Let's now open a new Jupyter notebook, and explore another dataset `GRCm38.gff3` from the `data/` folder.

[GFF is a standard file format](http://gmod.org/wiki/GFF3) for storing genomic features in a text file. GFF stands for Generic Feature Format. GFF files are plain text, 9 column, tab-delimited files. 

The 9 columns of the annotation section are as follows:

- Column 1: "seqid" - The ID of the landmark used to establish the coordinate system for the current feature.
- Column 2: "source" - The source is a free text qualifier intended to describe the algorithm or operating procedure that generated this feature. 
- Column 3: "type" - The type of the feature.
- Columns 4 & 5: "start" and "end" - The start and end of the feature.
- Column 6: "score" - The score of the feature, a floating point number.
- Column 7: "strand" - The strand of the feature.
- Column 8: "phase" - For features of type "CDS", the phase indicates where the feature begins with reference to the reading frame. 
- Column 9: "attributes" - A list of feature attributes in the format tag=value. 

We have modified these files and added a 10th column "gbid" which is the GenBank ID of each feature, and taken a random subset of these features for each species.

## Next session

Go to our next notebook: [Session 2.2: Data visualisation with Matplotlib](22_python_data.ipynb)